### PR TITLE
Share AssetFileDescriptor with ImageDecoder

### LINKED ISSUE #1970: [Coil 3] Lazy allocated BufferedSource for ImageSource
Now that we prefer ImageDecoderDecoder for API above 28
https://github.com/coil-kt/coil/pull/1969
ImageDecoder prefer a (File, Uri) not inputstream, we have no reason to preallocate a BufferedSource(this opens a native file descriptor), save this overhead since ImageDecoder will not use that fd
See https://github.com/VDostoyevskiy/coil/blob/068e566ce04e9e1d77e08b4c80c386870d876860/coil-core/src/androidMain/kotlin/coil3/fetch/AssetUriFetcher.kt#L25
https://github.com/VDostoyevskiy/coil/blob/068e566ce04e9e1d77e08b4c80c386870d876860/coil-core/src/androidMain/kotlin/coil3/fetch/ContentUriFetcher.kt#L33

### PR DESCRIPTION
**Pull request by VDostoyevskiy**
Close #1970

### CODE REVIEW

File: coil-core/src/androidMain/kotlin/coil3/decode/ImageSource.kt [hunk 1]
```diff
@@ -15,6 +16,7 @@ class AssetMetadata(
  */
 class ContentMetadata(
     val uri: Uri,
+    val afd: AssetFileDescriptor
```
Comment (colinrtwhite) [inline coil-core/src/androidMain/kotlin/coil3/decode/ImageSource.kt:19]:
```suggestion
    val assetFileDescriptor: AssetFileDescriptor,
```

Comment (colinrtwhite) [reply coil-core/src/androidMain/kotlin/coil3/decode/ImageSource.kt:19]: Should this be optional? Are there situations were an `AssetFileDescriptor` is not available for a URI?

File: coil-core/src/androidMain/kotlin/coil3/fetch/ContentUriFetcher.kt [hunk 2]
```diff
@@ -30,30 +30,28 @@ internal class ContentUriFetcher(
     override suspend fun fetch(): FetchResult {
         val androidUri = data.toAndroidUri()
         val contentResolver = options.context.contentResolver
-        val inputStream = if (isContactPhotoUri(data)) {
+        val afd = if (isContactPhotoUri(data)) {
```
Comment (colinrtwhite) [inline coil-core/src/androidMain/kotlin/coil3/fetch/ContentUriFetcher.kt:33]: Separate to this PR, but will `ImageDecoder.createSource(options.context.contentResolver, metadata.uri.toAndroidUri())` correctly handle this special case for a contacts image URI?

One of the issues with passing this data to `StaticImageDecoderDecoder` is we're coupling the source with the decoder.

File: coil-core/src/androidMain/kotlin/coil3/decode/ImageSource.kt (hunk 1 above)
Comment (VDostoyevskiy) [reply coil-core/src/androidMain/kotlin/coil3/decode/ImageSource.kt:19]: Nope, If we can open a InputStream for a ContentUri, there must have a afd as that I/OStream use afd internally

Comment (VDostoyevskiy) [reply coil-core/src/androidMain/kotlin/coil3/decode/ImageSource.kt]: Done

File: coil-core/src/androidMain/kotlin/coil3/fetch/ContentUriFetcher.kt (hunk 2 above)
Comment (VDostoyevskiy) [reply coil-core/src/androidMain/kotlin/coil3/fetch/ContentUriFetcher.kt:33]: > Separate to this PR, but will `ImageDecoder.createSource(options.context.contentResolver, metadata.uri.toAndroidUri())` correctly handle this special case for a contacts image URI?
> 
> One of the issues with passing this data to `StaticImageDecoderDecoder` is we're coupling the source with the decoder.

Haven't test it, but now ImageDecoder use a afd that InputStream use for this case so it doesn't matter

**Review by colinrtwhite — APPROVED** [review]

**Pull request merged**